src/puppetlabs/trapperkeeper/bootstrap.clj

```diff
@@ -132,30 +135,168 @@
+        (throw (bootstrap-error entry bootstrap-file line-number message))))))
+
+(schema/defn remove-duplicate-entries :- [AnnotatedBootstrapEntry]
+  "Removes any duplicate entries from the list of AnnotatedBootstrapEntry maps.
+   A entry is considered a duplicate if the :entry key is the same between two
+   entries.
+   Logs warnings for each duplicate found."
+  [entries :- [AnnotatedBootstrapEntry]]
+  (let [into-ordered-list
+        ; Accumulates bootstrap entries into :set to test for duplicates, and
+        ; accumulates the AnnotatedBootstrapEntry's into :ordered-entries so
+        ; that the order of their line numbers is maintained
+        (fn [acc annotated-entry]
+          (if (contains? (:set acc) (:entry annotated-entry))
+            (do (log/warn
+                 (format "Duplicate bootstrap entry found for service '%s' on line number '%s' in file '%s'"
+                         (:entry annotated-entry)
+                         (:line-number annotated-entry)
+                         (:bootstrap-file annotated-entry)))
+                acc)
+            {:set (conj (:set acc) (:entry annotated-entry))
+             :ordered-entries (conj (:ordered-entries acc) annotated-entry)}))]
+    ; Reduce the bootstrap entries into an ordered list of unique entries and return it
+    (:ordered-entries (reduce
+                       into-ordered-list
+                       {:set #{} :ordered-entries []}
+                       entries))))
 
-   In the case of an empty file, nil would be returned by line-seq, so we remove
-   all the nils at the end"
-  [files :- [(schema/protocol io/IOFactory)]]
-  (remove nil? (flatten (map #(line-seq (io/reader %)) files))))
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Public
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 
 (schema/defn parse-bootstrap-configs! :- [(schema/protocol services/ServiceDefinition)]
+  [configs :- [schema/Str]]
   "Parse multiple trapperkeeper bootstrap configuration files and return the
-  service graph that is the result of merging the graphs of all of the services
-  specified in the configuration files."
-  [configs :- [(schema/protocol io/IOFactory)]]
-  (let [lines (chain-files configs)]
-    (when (empty? lines) (throw (Exception. "Empty bootstrap config file")))
-    (for [line (map remove-comments lines)
-          :when (not (empty? line))]
-      (let [{:keys [namespace service-name]} (parse-bootstrap-line! line)]
-        (resolve-service! namespace service-name)))))
+  service graph that is the result of merging the graphs of all of the
+  services specified in the configuration files."
+  ; We remove the duplicate entries to allow the user to have duplicate entries in their
+  ; bootstrap files. If we didn't remove them, it would look like two services were trying
+  ; to implement the same protocol when we check for duplicate service implementations
```

Comment:
I think the rationale for this is that we'll probably have cross-over points where we want to move a bootstrap service entry from one file to another one - like with Puppet Server where we'll want to move the `certificate-authority` service entry from the main `bootstrap.cfg` file over to the new CA-specific `bootstrap.cfg` one (or whatever we end up calling it) - and we didn't necessarily want to cause the app to fail to launch just because we happened to have references to exactly the same service namespace in two different files - like one for the new file and one from the old file.  This would also seem to support the use cases being discussed in ![SERVER-1213](https://tickets.puppetlabs.com/browse/SERVER-1213) where some users might want to prepare for an upgrade early by laying down file content as they eventually want to have it after the upgrade living alongside the content as it needs to exist in order to work with the old release still being used - i.e., the monolithic bootstrap.cfg that you have to have today.  @cprice404 may want to provide more color on this when he gets back.  In any event, this behavior would at least be backward compatible with the current Trapperkeeper implementation.